File: sdk/ruby/lib/anyvali/schema.rb
# frozen_string_literal: true

module AnyVali
  class Schema
    attr_reader :kind, :constraints, :coerce_config, :default_value, :has_default,
                :custom_validators, :metadata

    RESERVED_METADATA_KEYS = %w[title description deprecated deprecatedMessage notStable since sensitive readonly writeonly examples].freeze

    def initialize(kind:, constraints: {}, coerce_config: nil, default_value: nil, has_default: false, custom_validators: [], metadata: {})
      @kind = kind
      @constraints = constraints.freeze
      @coerce_config = coerce_config
      @default_value = default_value
      @has_default = has_default
      @custom_validators = custom_validators.freeze
      @metadata = metadata.freeze
    end

    def parse(input)
      result = safe_parse(input)
      raise ValidationError, result.issues if result.failure?
      result.value
    end

    def safe_parse(input, path: [], context: nil)
      context ||= ValidationContext.new
      value = input
      issues = []

      # Step 1: Coercion (if present and configured)
      if @coerce_config && !value.nil?
        coerced = Coercion.apply(value, @coerce_config, @kind)
        if coerced[:success]
          value = coerced[:value]
        else
          issues << ValidationIssue.new(
            code: IssueCodes::COERCION_FAILED,
            path: path,
            expected: @kind,
            received: value.is_a?(String) ? value : value.to_s
          )
          return ParseResult.new(value: nil, issues: issues)
        end
      end

      # Step 2: Validate
      validate(value, path, issues, context)

      # Step 3: Custom validators
      if issues.empty? && !@custom_validators.empty?
        @custom_validators.each do |validator|
          validator_issues = validator.call(value, path)
          issues.concat(validator_issues) if validator_issues
        end
      end

      if issues.empty?
        ParseResult.new(value: value, issues: [])
      else
        ParseResult.new(value: nil, issues: issues)
      end
    end

    def default(value)
      dup_with(default_value: value, has_default: true)
    end

    # Enable coercion on this schema.
    #
    # With no argument (`schema.coerce`) the only portable coercion source —
    # "string" (spec 5.1) — is implied, and the coercion target is inferred
    # from the schema kind (e.g. number().coerce coerces a string to a number).
    # An explicit config string/array (e.g. "string->int", "trim",
    # ["trim", "lower"]) is still accepted for backwards compatibility.
    def coerce(config = "string")
      dup_with(coerce_config: config)
    end

    def describe(description, **opts)
      raise ArgumentError, "describe(): description must be a string" unless description.is_a?(String)

      meta = { "description" => description }

      if opts.key?(:title)
        raise ArgumentError, "describe(): title must be a string" unless opts[:title].is_a?(String)
        meta["title"] = opts[:title]
      end
      if opts.key?(:deprecated)
        raise ArgumentError, "describe(): deprecated must be a boolean" unless [true, false].include?(opts[:deprecated])
        meta["deprecated"] = opts[:deprecated]
      end
      if opts.key?(:deprecated_message)
        raise ArgumentError, "describe(): deprecatedMessage must be a string" unless opts[:deprecated_message].is_a?(String)
        raise ArgumentError, "describe(): deprecatedMessage requires deprecated: true" unless opts[:deprecated]
        meta["deprecatedMessage"] = opts[:deprecated_message]
      end
      if opts.key?(:not_stable)
        raise ArgumentError, "describe(): notStable must be a boolean" unless [true, false].include?(opts[:not_stable])
        meta["notStable"] = opts[:not_stable]
      end
      if opts.key?(:since)
        raise ArgumentError, "describe(): since must be a string" unless opts[:since].is_a?(String)
        meta["since"] = opts[:since]
      end
      if opts.key?(:sensitive)
        raise ArgumentError, "describe(): sensitive must be a boolean" unless [true, false].include?(opts[:sensitive])
        meta["sensitive"] = opts[:sensitive]
      end
      if opts.key?(:readonly)
        raise ArgumentError, "describe(): readonly must be a boolean" unless [true, false].include?(opts[:readonly])
        meta["readonly"] = opts[:readonly]
      end
      if opts.key?(:writeonly)
        raise ArgumentError, "describe(): writeonly must be a boolean" unless [true, false].include?(opts[:writeonly])
        meta["writeonly"] = opts[:writeonly]
      end
      if opts[:readonly] && opts[:writeonly]
        raise ArgumentError, "describe(): readonly and writeonly cannot both be true"
      end
      if opts.key?(:examples)
        raise ArgumentError, "describe(): examples must be an array" unless opts[:examples].is_a?(Array)
        meta["examples"] = opts[:examples]
      end

      dup_with(metadata: @metadata.merge(meta))
    end

    def with_metadata(meta, replace: false)
      meta.each_key do |key|
        if RESERVED_METADATA_KEYS.include?(key)
          raise ArgumentError, "with_metadata(): \"#{key}\" is a reserved key. Use describe() instead."
        end
      end

      if replace
        preserved = @metadata.select { |k, _| RESERVED_METADATA_KEYS.include?(k) }
        dup_with(metadata: preserved.merge(meta))
      else
        dup_with(metadata: @metadata.merge(meta))
      end
    end

    def refine(&block)
      dup_with(custom_validators: @custom_validators + [block])
    end

    def portable?
      @custom_validators.empty?
    end

    def export(mode: :portable)
      if mode == :portable && !portable?
        raise ValidationError, [
          ValidationIssue.new(
            code: IssueCodes::CUSTOM_VALIDATION_NOT_PORTABLE,
            expected: "portable schema",
            received: "schema with custom validators"
          )
        ]
      end
      doc = AnyValiDocument.new(root: self)
      doc.to_h
    end

    def to_node
      node = { "kind" => @kind }
      @constraints.each { |k, v| node[k] = v }
      node["coerce"] = @coerce_config if @coerce_config
      node["default"] = @default_value if @has_default
      node["metadata"] = @metadata.dup unless @metadata.empty?
      node
    end

    protected

    def validate(value, path, issues, context)
      raise NotImplementedError, "Subclasses must implement validate"
    end

    def dup_with(**overrides)
      attrs = {
        kind: @kind,
        constraints: @constraints,
        coerce_config: @coerce_config,
        default_value: @default_value,
        has_default: @has_default,
        custom_validators: @custom_validators,
        metadata: @metadata
      }.merge(overrides)
      self.class.new(**attrs)
    end

    # Helper to get the JSON type name for a Ruby value
    def self.type_name(value)
      case value
      when NilClass then "null"
      when TrueClass, FalseClass then "boolean"
      when Integer then "number"
      when Float then "number"
      when String then "string"
      when Array then "array"
      when Hash then "object"
      else value.class.name.downcase
      end
    end
  end
end

File: sdk/csharp/src/AnyVali/Schema.cs
using AnyVali.Parse;

namespace AnyVali;

/// <summary>
/// Abstract base for all AnyVali schemas.
/// </summary>
public abstract class Schema
{
    internal const string AnyvaliVersionValue = "1.0";
    internal const string SchemaVersionValue = "1";

    internal object? DefaultValue { get; set; } = Absent.Value;
    internal CoercionConfig? CoercionCfg { get; set; }
    internal bool IsPortable { get; set; } = true;

    // ---- Public API ----

    /// <summary>
    /// Parse input. Throws ValidationError on failure.
    /// </summary>
    public object? Parse(object? input)
    {
        var result = SafeParse(input);
        if (result.Success) return result.Data;
        throw new ValidationError(result.Issues);
    }

    /// <summary>
    /// Parse input. Returns ParseResult with success/failure.
    /// </summary>
    public ParseResult SafeParse(object? input)
    {
        var ctx = new ValidationContext();
        var output = RunPipeline(input, ctx);
        if (ctx.Issues.Count > 0)
            return ParseResult.Fail(ctx.Issues);
        return ParseResult.Ok(output);
    }

    /// <summary>
    /// Run the 5-step parse pipeline.
    /// </summary>
    internal virtual object? RunPipeline(object? input, ValidationContext ctx)
    {
        // Step 1: detect presence
        var isAbsent = Absent.IsAbsent(input) || input is null && IsAbsentNull(input);

        var value = input;

        // Step 2: coercion (only for present values)
        if (!isAbsent && CoercionCfg is not null)
        {
            var coerced = Coercion.Apply(value, CoercionCfg, GetCoercionTarget());
            if (coerced.Success)
            {
                value = coerced.Value;
            }
            else
            {
                ctx.Issues.Add(new ValidationIssue
                {
                    Code = IssueCodes.CoercionFailed,
                    Message = coerced.Message ?? "Coercion failed",
                    Path = ctx.ClonePath(),
                    Expected = GetCoercionTarget(),
                    Received = DescribeValue(input),
                });
                return null;
            }
        }

        // Step 3: default materialization (only for absent values)
        var usedDefault = false;
        if (isAbsent && !Absent.IsAbsent(DefaultValue))
        {
            value = DefaultValue;
            usedDefault = true;
        }

        // Step 4: validate
        var issuesBefore = ctx.Issues.Count;
        var result = Validate(value, ctx);

        // If default was materialized and validation failed, remap to default_invalid
        if (usedDefault && ctx.Issues.Count > issuesBefore)
        {
            for (var i = issuesBefore; i < ctx.Issues.Count; i++)
            {
                var issue = ctx.Issues[i];
                ctx.Issues[i] = new ValidationIssue
                {
                    Code = IssueCodes.DefaultInvalid,
                    Message = issue.Message,
                    Path = issue.Path,
                    Expected = issue.Expected,
                    Received = issue.Received,
                    Meta = issue.Meta,
                };
            }
        }

        return result;
    }

    /// <summary>
    /// Whether a null input should be treated as absent.
    /// By default, null is NOT absent (it's a concrete null value).
    /// Only overridden in contexts where null means "missing".
    /// </summary>
    protected virtual bool IsAbsentNull(object? input) => false;

    /// <summary>
    /// Override to provide the coercion target type name.
    /// </summary>
    internal virtual string GetCoercionTarget() => "unknown";

    /// <summary>
    /// Core validation logic. Override in subclasses.
    /// </summary>
    internal abstract object? Validate(object? input, ValidationContext ctx);

    /// <summary>
    /// Convert to interchange schema node.
    /// </summary>
    internal abstract Dictionary<string, object?> ToNode();

    /// <summary>
    /// Set a default value. Returns a clone.
    /// </summary>
    public Schema Default(object? value)
    {
        var clone = Clone();
        clone.DefaultValue = value;
        return clone;
    }

    /// <summary>
    /// Enable coercion. Returns a clone.
    /// </summary>
    public Schema Coerce(CoercionConfig? config = null)
    {
        var clone = Clone();
        clone.CoercionCfg = config ?? new CoercionConfig();
        return clone;
    }

    /// <summary>
    /// Export to AnyValiDocument.
    /// </summary>
    public AnyValiDocument Export(ExportMode mode = ExportMode.Portable)
    {
        if (mode == ExportMode.Portable && !IsPortable)
            throw new InvalidOperationException(
                "Cannot export in portable mode: schema contains non-portable features");

        var node = ToNode();
        return new AnyValiDocument
        {
            AnyvaliVersion = AnyvaliVersionValue,
            SchemaVersion = SchemaVersionValue,
            Root = node,
            Definitions = new Dictionary<string, object?>(),
            Extensions = new Dictionary<string, object?>(),
        };
    }

    // ---- Internal helpers ----

    internal abstract Schema Clone();

    internal void AddDefaultAndCoercion(Dictionary<string, object?> node)
    {
        if (!Absent.IsAbsent(DefaultValue))
            node["default"] = DefaultValue;
        if (CoercionCfg is not null)
        {
            var coerce = new Dictionary<string, object?>();
            if (CoercionCfg.From is not null) coerce["from"] = CoercionCfg.From;
            if (CoercionCfg.Trim) coerce["trim"] = true;
            if (CoercionCfg.Lower) coerce["lower"] = true;
            if (CoercionCfg.Upper) coerce["upper"] = true;
            node["coerce"] = coerce;
        }
    }

    internal static string DescribeType(object? value)
    {
        if (value is null) return "null";
        if (Absent.IsAbsent(value)) return "undefined";
        if (value is bool) return "boolean";
        if (value is string) return "string";
        if (value is long or int or short or byte or sbyte or ushort or uint or ulong) return "number";
        if (value is double or float or decimal) return "number";
        if (value is List<object?> or object[]) return "array";
        if (value is Dictionary<string, object?>) return "object";
        return value.GetType().Name.ToLowerInvariant();
    }

    internal static string DescribeValue(object? value)
    {
        if (value is null) return "null";
        if (value is string s) return s;
        return value.ToString() ?? "null";
    }

    /// <summary>
    /// Convert a numeric value to double for comparison.
    /// </summary>
    internal static double ToDouble(object? value)
    {
        return value switch
        {
            double d => d,
            long l => l,
            int i => i,
            float f => f,
            decimal m => (double)m,
            short s => s,
            byte b => b,
            sbyte sb => sb,
            ushort us => us,
            uint ui => ui,
            ulong ul => ul,
            _ => double.NaN
        };
    }

    /// <summary>
    /// Check if a value is a finite number.
    /// </summary>
    internal static bool IsFiniteNumber(object? value)
    {
        if (value is double d) return double.IsFinite(d);
        if (value is float f) return float.IsFinite(f);
        return value is long or int or short or byte or sbyte or ushort or uint or ulong or decimal;
    }

    /// <summary>
    /// Check if a value is an integer (no fractional part).
    /// </summary>
    internal static bool IsInteger(object? value)
    {
        if (value is long or int or short or byte or sbyte or ushort or uint or ulong) return true;
        if (value is double d) return double.IsFinite(d) && d == Math.Floor(d);
        if (value is float f) return float.IsFinite(f) && f == MathF.Floor(f);
        if (value is decimal m) return m == Math.Floor(m);
        return false;
    }

    /// <summary>
    /// Convert a numeric value to long.
    /// </summary>
    internal static long ToLong(object? value)
    {
        return value switch
        {
            long l => l,
            int i => i,
            double d => (long)d,
            float f => (long)f,
            decimal m => (long)m,
            short s => s,
            byte b => b,
            sbyte sb => sb,
            ushort us => us,
            uint ui => ui,
            ulong ul => (long)ul,
            _ => 0
        };
    }
}

/// <summary>
/// Generic base for schemas that provides typed Parse and SafeParse methods.
/// </summary>
public abstract class Schema<T> : Schema
{
    /// <summary>
    /// Parse input with strong typing. Throws ValidationError on failure.
    /// </summary>
    public new T Parse(object? input)
    {
        var result = base.SafeParse(input);
        if (result.Success) return CastData(result.Data)!;
        throw new ValidationError(result.Issues);
    }

    /// <summary>
    /// Parse input with strong typing. Returns ParseResult&lt;T&gt; with success/failure.
    /// </summary>
    public ParseResult<T> SafeParseTyped(object? input)
    {
        var result = base.SafeParse(input);
        if (result.Success)
            return ParseResult<T>.Ok(CastData(result.Data));
        return ParseResult<T>.Fail(result.Issues);
    }

    private static T? CastData(object? data)
    {
        if (data is null) return default;
        if (data is T typed) return typed;
        // Handle numeric type conversions (e.g., long -> double for IntSchema via NumberSchema)
        return (T)Convert.ChangeType(data, Nullable.GetUnderlyingType(typeof(T)) ?? typeof(T));
    }
}
